File: Android/AmigoSurvey/app/src/main/java/com/amigocloud/amigosurvey/form/AmigoBridge.java
/*
 *
 *  AmigoMobile
 *
 *  Copyright (c) 2011-2015 AmigoCloud Inc., All rights reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 3.0 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this library.
 *
 */



package com.amigocloud.amigomobile.api.js;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.webkit.JavascriptInterface;

import com.amigocloud.amigosurvey.form.FormActivity;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class AmigoBridge {


	public static final int REQUEST_IMAGE_CAPTURE = 1;
	public static final int REQUEST_VIDEO_CAPTURE = 2;
	public static final int REQUEST_PHOTO_GALLERY = 100;

	private Activity activity;
	private FormActivity formActivity;
	private String formType;
	private String formData;
	private long datasetId;
	private boolean deleteRelationship;
    private String amigoId;

	public AmigoBridge(Activity activity, FormActivity formActivity, long datasetId, String data) {
		this.activity = activity;
		this.formActivity = formActivity;
		this.formData = data;
		this.datasetId = datasetId;
		if (datasetId < 0) {
			IllegalStateException e =  new IllegalStateException("Dataset ID cannot be < 0");
//			Crashlytics.logException(e);
			throw e;
		}
	}

	public void setFormType(String type) {
		formType = type;
	}

    public String getFormType() {
        return formType;
    }

    public long getDatasetId() {
        return datasetId;
    }

    @JavascriptInterface
    public String getDatasetInfo(String datasetId) {
        return ""; // TODO: AmigoCloudAPI.getDatasetInfo(Long.parseLong(datasetId));
    }

    public void setCustomFieldValue(final String fieldName, final String fieldValue) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (formActivity.isReady())
                    formActivity.getWebView().loadUrl("javascript:Amigo.setCustomFieldValue(" +
                            String.valueOf(datasetId) + ",'" + fieldName + "','" + fieldValue + "')");
            }
        });
    }

    @JavascriptInterface
    public String getClientType() {
        return "mobile";
    }

    @JavascriptInterface
    public void ready() {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (formActivity.isReady())
                    formActivity.getWebView().loadUrl("javascript:Amigo.loadBlock(" +
                            "AmigoPlatform.getBlockHTML('" + formType + "'), '" + formType + "', " +
                            "'" + datasetId + "', AmigoPlatform.getData());");
            }
        });
    }

    @JavascriptInterface
    public String getData() {
        return formData;
    }

    @JavascriptInterface
    public int getPageSize() {
        return 20;
    }

    public void historyBack() {
        if (formActivity.isReady()) {
            formActivity.getWebView().loadUrl("javascript:Amigo.historyBack()");
        }
    }

    public void mediaAdded() {
        if (formActivity.isReady()) {
            formActivity.getWebView().loadUrl("javascript:Amigo.mediaAdded()");
        }
    }

    @JavascriptInterface
    public void onException(String msg) {
        // Do nothing, will be deprecated
    }

    @JavascriptInterface
    public void storeException(String url, String data) {
//   TODO:     AmigoCloudAPI.storeException(url, data);
    }

    @JavascriptInterface
    public void openUrl(final String url) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                activity.startActivity(browserIntent);
            }
        });
    }

    public String getAmigoId() {
        return amigoId;
    }

    private void extractAmigoId(String json) {
        try {
            JSONObject jObject = new JSONObject(json);
            String data = jObject.getString("data");
            JSONObject jObject2 = new JSONObject(data);
            JSONArray dataArray = jObject2.getJSONArray("data");
            if (dataArray.length() == 1) {
                amigoId = ((JSONObject) dataArray.get(0)).getString("amigo_id");
            } else
                amigoId = "";
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @JavascriptInterface
    public void setState(String json) {
//        AmigoCloudAPI.clearEditorContext();
        extractAmigoId(json);
        try {
            JSONObject jObject = new JSONObject(json);
            formType = jObject.getString("formType");
            formData = jObject.getString("data");
            datasetId = Long.parseLong(jObject.getString("currentDatasetId"));
            final boolean saveButton = jObject.getBoolean("saveBtn");
            final boolean deleteButton = jObject.getBoolean("deleteBtn");
            final boolean checkButton = jObject.getBoolean("checkBtn");
            deleteRelationship = jObject.getBoolean("relationshipDeleteBtn");
// TODO:
//            activity.runOnUiThread(new Runnable() {
//                @Override
//                public void run() {
//                    if (formActivity instanceof BaseFormFragment) {
//                        BaseFormFragment formFragment = (BaseFormFragment) formActivity;
//                        formFragment.setSaveButtonEnabled(saveButton);
//                        formFragment.setDeleteButtonEnabled(deleteButton);
//                        formFragment.setCheckButtonEnabled(checkButton);
//                        formFragment.setHeaderTitle(AmigoCloudAPI.getDatasetName(datasetId));
//                    }
//                }
//            });
        } catch (NumberFormatException | JSONException e) {
//            Crashlytics.logException(e);
            e.printStackTrace();
        }

    }

    @JavascriptInterface
    public String getFormDescription(String datasetId, String formType) {
        String type = formType + "_description";
        return ""; // TODO: AmigoCloudAPI.getDatasetForm(Long.valueOf(datasetId), type);
    }

    @JavascriptInterface
    public String getBlockHTML(String formType) {
        return ""; // TODO: AmigoCloudAPI.getDatasetForm(datasetId, formType);
    }

    @JavascriptInterface
    public String getBlockHTMLWithId(String formType, String datasetId) {
        try {
            return ""; // TODO: AmigoCloudAPI.getDatasetForm(Long.parseLong(datasetId), formType);
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getDatasetRecords(String datasetId, String offset) {
        return getDatasetRecordsWithFilter(datasetId, offset, "");
    }

    @JavascriptInterface
    public String getDatasetRecordsWithFilter(String datasetId, String offset, String filter) {
        try {
            return ""; // TODO: AmigoCloudAPI.getRecordsJSON(Long.parseLong(datasetId), Integer.parseInt(offset), filter, "");
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getDatasetRecordsDistinctColumnsWithFilter(String datasetId, String columns, String filter) {
        long dsid = 0;
        try {
            dsid = Long.parseLong(datasetId);
        } catch (NumberFormatException e) {
            e.printStackTrace();
            return "{}";
        }
        return ""; // TODO: AmigoCloudAPI.getDistinctRecordsJSON(dsid, 0, filter, "", columns);
    }

    @JavascriptInterface
    public String getDatasetRecordsWithFilterOrderBy(String datasetId, String offset, String filter, String order_by) {
        try {
            return ""; // TODO: AmigoCloudAPI.getRecordsJSON(Long.parseLong(datasetId), Integer.parseInt(offset), filter, order_by);
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getDatasetRecordsCountWithId(String datasetId) {
        return getDatasetRecordsCountWithIdAndFilter(datasetId, "");
    }

    @JavascriptInterface
    public String getDatasetRecordsCountWithIdAndFilter(String datasetId, String filter) {
        try {
            return ""; // TODO: AmigoCloudAPI.getRecordsCountJSON(Long.parseLong(datasetId), filter);
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getRelatedRecordsCount(String originalDatasetId, String relatedDatasetId, String sourceAmigoId) {
        try {
            return ""; // TODO: AmigoCloudAPI.getRelatedRecordsCountJSON(Long.parseLong(originalDatasetId), Long.parseLong(relatedDatasetId), sourceAmigoId);
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getNewRowWithSourceId(String datasetId, String relatedColumn, String relatedId) {
        try {
            return ""; // TODO: AmigoCloudAPI.getNewRecordJSONWithRelationship(Long.parseLong(datasetId), relatedColumn, relatedId);
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getNewRow(String datasetId) {
        try {
            return ""; // TODO: AmigoCloudAPI.getNewRecordJSONForDataset(Long.parseLong(datasetId));
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getSchemaWithId(String datasetId) {
        try {
            return ""; // TODO: AmigoCloudAPI.getSchemaJSONForDataset(Long.parseLong(datasetId));
        } catch (NumberFormatException e) {
            return "";
        }
    }

    /**
     * Old Form Methods
     **/

    @JavascriptInterface
    public String getSchema() {
        return ""; // TODO: AmigoCloudAPI.getSchemaJSONForDataset(datasetId);
    }

    @JavascriptInterface
    public String getUser() {
        return ""; // TODO: AmigoCloudAPI.getUserInfoJSON();
    }

    @JavascriptInterface
    public String getProject() {
        return ""; // TODO: AmigoCloudAPI.getSelectedProjectInfoJSON();
    }

    @JavascriptInterface
    public String getRelatedTables(String datasetId) {
        try {
            return ""; // TODO: AmigoCloudAPI.getRelatedTablesJSON(Long.parseLong(datasetId));
        } catch (NumberFormatException e) {
            return "";
        }
    }

    @JavascriptInterface
    public String getGPSinfo() {
        return ""; // TODO: Globe.getGPSInfoJSON();
    }

    @JavascriptInterface
    public String getPermissionLevel() {
        String permission = "EDITOR"; // TODO: AmigoCloudAPI.getProjectPermissionLevel(AmigoCloudAPI.getSelectedProjectId());
        if (permission == null || permission.isEmpty()) {
            return "EDITOR";
        }
        return permission;
    }

    @JavascriptInterface
    public void editRowGeometry(final String json) {
//        TODO:
//        AmigoApplication.Companion.sendAnalyticsEvent(R.string.category_user, R.string.action_edit_geometry);
//        GeometryEditorParams.Companion.setParams(new GeometryEditorParams(datasetId, json));
//        if (activity instanceof GlobeActivity) {
//            GlobeActivityEditExtensionsKt.enableEditMode((GlobeActivity) activity, formType.equals("create_block"), null);
//        }
    }

    @JavascriptInterface
    public void updateRow(final String originalJson, final String updateJson) {
//        TODO:
//        if (formType.equals("create_block"))
//            AmigoApplication.Companion.sendAnalyticsEvent(R.string.category_user, R.string.action_create_record);
//        else
//            AmigoApplication.Companion.sendAnalyticsEvent(R.string.category_user, R.string.action_change_record);
//
//        AmigoCloudAPI.addToStagingData(datasetId, originalJson, updateJson);
//
//        if (activity instanceof GlobeActivity) {
//            GlobeActivityEditExtensionsKt.disableEditMode((GlobeActivity) activity);
//        }
    }

    @JavascriptInterface
    public void saveRow(final String json) {
        updateRow(formData, json);
    }

    @JavascriptInterface
    public void close() {
//        TODO:
//        if (AmigoCloudAPI.flushStagingData()) {
//            if(AmigoCloudAPI.isAutoSyncEnabled()) {
//                AmigoCloudAPI.syncProject(AmigoCloudAPI.getSelectedProjectId());
//            }
//            AmigoCloudAPI.clearDrawingCache();
//        } else {
//            activity.runOnUiThread(new Runnable() {
//                @Override
//                public void run() {
//                    AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(activity);
//                    alertDialogBuilder.setTitle(activity.getString(R.string.error));
//                    alertDialogBuilder.setMessage(activity.getString(R.string.malformed_data));
//                    alertDialogBuilder.setNegativeButton(activity.getString(R.string.ui_ok), null);
//                    alertDialogBuilder.create().show();
//                }
//            });
//        }
//        activity.runOnUiThread(new Runnable() {
//            @Override
//            public void run() {
//                if (activity instanceof GlobeActivity) {
//                    GlobeActivityFragmentExtensionsKt.clearPopUps(((GlobeActivity) activity), GlobeActivity.Companion.getCLEAR_ALL());
//                } else
//                    activity.onBackPressed();
//            }
//        });
    }

	@JavascriptInterface
	public void deleteRows(final String json) {
//        TODO:
//		activity.runOnUiThread(new Runnable() {
//			@Override
//			public void run() {
//				AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(activity);
//				if(deleteRelationship) {
//					alertDialogBuilder.setTitle(activity.getString(R.string.delete_rows));
//					alertDialogBuilder.setMessage(activity.getString(R.string.delete_relationships));
//					alertDialogBuilder.setPositiveButton(activity.getString(R.string.relationships_only), new DialogInterface.OnClickListener() {
//						public void onClick(DialogInterface dialog, int id) {
//							if (formActivity.isReady()) {
//								formActivity.getWebView().loadUrl("javascript:Amigo.deleteSelectedRelationships()");
//							}
//							dialog.dismiss();
//						}
//					});
//					alertDialogBuilder.setNeutralButton(activity.getString(R.string.ui_both), new DialogInterface.OnClickListener() {
//						public void onClick(DialogInterface dialog, int id) {
//							AmigoApplication.Companion.sendAnalyticsEvent(R.string.category_user, R.string.action_delete_record);
//							AmigoCloudAPI.deleteRows(datasetId, json);
//							historyBack();
//							dialog.dismiss();
//						}
//					});
//					alertDialogBuilder.setNegativeButton(activity.getString(R.string.ui_cancel), null);
//				} else {
//					alertDialogBuilder.setTitle(activity.getString(R.string.delete_rows));
//					alertDialogBuilder.setMessage(activity.getString(R.string.are_you_sure));
//					alertDialogBuilder.setPositiveButton(activity.getString(R.string.ui_yes), new DialogInterface.OnClickListener() {
//						public void onClick(DialogInterface dialog, int id) {
//							AmigoApplication.Companion.sendAnalyticsEvent(R.string.category_user, R.string.action_delete_record);
//							AmigoCloudAPI.deleteRows(datasetId, json);
//							historyBack();
//							dialog.dismiss();
//						}
//					});
//					alertDialogBuilder.setNegativeButton(activity.getString(R.string.ui_no), null);
//				}
//				alertDialogBuilder.create().show();
//			}
//		});
	}

	@JavascriptInterface
	public void newRecord() {
//        TODO:
//		String newRecordData = AmigoCloudAPI.getNewRecordJSONForDataset(datasetId);
//		if(activity instanceof GlobeActivity) {
//            GlobeActivityFragmentExtensionsKt.showEditForm((GlobeActivity) activity, datasetId, newRecordData, true);
//		}
	}

	@JavascriptInterface
	public void zoomToRows(final String json) {
//        TODO:
//		AmigoApplication.Companion.sendAnalyticsEvent(R.string.category_user, R.string.action_zoom_to_record);
//		AmigoCloudAPI.showDataset(datasetId);
//		AmigoCloudAPI.zoomToRecords(datasetId, json);
	}

	@JavascriptInterface
	public void takePhoto(final String relatedTableId, final String amigoId) {
//        TODO:
//		if(activity instanceof GlobeActivity) {
//            final GlobeActivity globeActivity = (GlobeActivity) activity;
//            AlertDialog.Builder builder = new AlertDialog.Builder(activity);
//            String[] entries = {activity.getString(R.string.take_photo),
//                    activity.getString(R.string.capture_video),
//                    activity.getString(R.string.use_existing_file)};
//            builder.setItems(entries, new DialogInterface.OnClickListener() {
//                @Override
//                public void onClick(DialogInterface dialogInterface, int which) {
//                    if (which == 0 || which == 1) {
//                        Intent takePictureIntent;
//                        String fileExtension;
//                        int mode = 0;
//                        if (which == 0) {
//                            fileExtension = ".jpg";
//                            takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
//                            mode = REQUEST_IMAGE_CAPTURE;
//                        } else {
//                            fileExtension = ".mp4";
//                            takePictureIntent = new Intent(MediaStore.ACTION_VIDEO_CAPTURE);
//                            mode = REQUEST_VIDEO_CAPTURE;
//                        }
//                        String storageDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES) + File.separator + "amigocloud";
//                        File storageDirFile = new File(storageDir);
//                        if (!storageDirFile.exists())
//                            storageDirFile.mkdirs();
//                        String timeStamp = new SimpleDateFormat("yyyyMMddHHmmssSSS", Locale.US).format(new Date());
//
//                        File image = new File(storageDir, relatedTableId + timeStamp + fileExtension);
//
//                        globeActivity.setPhotoInfo(new PhotoInfo(datasetId, Long.parseLong(relatedTableId), amigoId, image));
//                        takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(image));
//                        try {
//                            activity.startActivityForResult(takePictureIntent, mode);
//                        } catch (Exception e) {
//                            Crashlytics.logException(e);
//                            e.printStackTrace();
//                        }
//                    } else {
//                        globeActivity.setPhotoInfo(new PhotoInfo(datasetId, Long.parseLong(relatedTableId), amigoId, null));
//                        Intent photoPickerIntent = new Intent(Intent.ACTION_GET_CONTENT, android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
//                        photoPickerIntent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
//                        photoPickerIntent.setType("image/* video/* audio/* text/* application/*");
//                        try {
//                            activity.startActivityForResult(photoPickerIntent, REQUEST_PHOTO_GALLERY);
//                        } catch (Exception e) {
//                            Crashlytics.logException(e);
//                            e.printStackTrace();
//                        }
//                    }
//
//                    dialogInterface.dismiss();
//                }
//            });
//            builder.create().show();
//        }
	}

    @JavascriptInterface
    public void scanBarcode(String amigoId) {
//        TODO:
//        IntentIntegrator integrator = new IntentIntegrator(activity);
//        final GlobeActivity globeActivity = (GlobeActivity) activity;
//        globeActivity.setPhotoInfo(new PhotoInfo(datasetId, 0, amigoId, null));
//        integrator.initiateScan();
    }

	@JavascriptInterface
    public void viewPhotos(String relatedTableId, String amigoId) {
//        TODO:
//        if (activity instanceof GlobeActivity) {
//            GlobeActivityFragmentExtensionsKt.showPhotoGridFragment((GlobeActivity) activity,
//                    AmigoCloudAPI.getFileArray(datasetId, Long.parseLong(relatedTableId), amigoId),
//                    datasetId, Long.parseLong(relatedTableId), amigoId);
//        }
	}

	@JavascriptInterface
	public void dataHasChanged(final boolean hasChanged) {
//        TODO:
//		activity.runOnUiThread(new Runnable() {
//			@Override
//			public void run() {
//				if(hasChanged) {
//					AlertDialog.Builder ad = new AlertDialog.Builder(activity, R.style.AmigoThemeDialog);
//					ad.setTitle(activity.getString(R.string.warning));
//					ad.setMessage(activity.getString(R.string.discard_changes));
//
//					ad.setNegativeButton(activity.getString(R.string.ui_cancel), null);
//					ad.setPositiveButton(activity.getString(R.string.ui_discard), new DialogInterface.OnClickListener() {
//						@Override
//						public void onClick(DialogInterface dialogInterface, int i) {
//                            AmigoCloudAPI.cleanStagingData();
//                            historyBack();
//                            dialogInterface.dismiss();
//                        }
//                    });
//                    ad.show();
//                } else {
//                    historyBack();
//                }
//            }
//        });
    }

    @JavascriptInterface
    public void writeRfidInfo(String lat, String lon, String owner, String asset_name) {
        if(lat==null) lat = "0.0";
        if(lon==null) lon = "0.0";
        if(owner==null) owner = "N/A";
        if(asset_name==null) asset_name = "N/A";
        final String latf = lat;
        final String lonf = lon;
        final String ownerf = owner;
        final String asset_namef = asset_name;

//        TODO:
//        if (activity instanceof GlobeActivity) {
//            final GlobeActivity globeActivity = (GlobeActivity) activity;
//            GlobeActivityTSLExtensionsKt.writeToTSLMarker(globeActivity, amigoId, datasetId, latf, lonf, ownerf, asset_namef);
//        }
    }

    @JavascriptInterface
    public String getGeometryInfo(String wkb) {
        return ""; // TODO: AmigoCloudAPI.getGeometryInfo(wkb);
    }

    public void setRfidWriteStatus(final String text, final String status) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (formActivity.isReady())
                    formActivity.getWebView().loadUrl("javascript:Amigo.setRfidWriteStatus('" +
                            text + "','" + status + "')");
            }
        });
    }
}

File: Android/AmigoSurvey/app/src/main/java/com/amigocloud/amigosurvey/form/AmigoBridge.kt
/*
 *
 *  AmigoMobile
 *
 *  Copyright (c) 2011-2015 AmigoCloud Inc., All rights reserved.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 3.0 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this library.
 *
 */

package com.amigocloud.amigosurvey.form

import android.webkit.JavascriptInterface
import android.content.Intent
import android.location.Location
import android.net.Uri



class AmigoBridge(private val formActivity: FormActivity) {

    var formType: String = ""

    var formViewState: FormViewState = FormViewState()

    var lastLocation: Location = Location("")
    var geometrySet = false

    fun getLastLocationWKT() : String {
        return "SRID=4326;POINT(${lastLocation.longitude} ${lastLocation.latitude})"
    }

    fun setGeometryPosition() {
        refreshGeometry(getLastLocationWKT())
        geometrySet = true
    }

    @JavascriptInterface
    fun getClientType(): String {
        return "mobile"
    }

    @JavascriptInterface
    fun ready() {
        runJS("javascript:Amigo.loadBlock(" +
                        "AmigoPlatform.getBlockHTML('" + formType + "'), '" + formType + "', " +
                        "'" + formViewState.dataset?.id + "', AmigoPlatform.getData());")
    }

    fun runJS(url: String) {
        formActivity.runOnUiThread(Runnable {
            if (formActivity.isReady())
                formActivity.getWebView().loadUrl(url)
        })
    }

    fun back()
    {
        runJS("javascript:Amigo.historyBack();")
    }

    fun refreshGeometry(wkt: String) {
        runJS("javascript:Amigo.refreshGeometry('$wkt');")
    }

    fun submit() {
        if (!geometrySet) {
            setGeometryPosition()
        }
        runJS("javascript:Amigo.nativeSaveButton();")
    }

    @JavascriptInterface
    fun getData(): String {
        return formViewState.recordJson
    }

    @JavascriptInterface
    fun getPageSize(): Int {
        return 20
    }

    @JavascriptInterface
    fun onException(msg: String) {
        // Do nothing, will be deprecated
        print("onException() -----: " + msg)
    }

    @JavascriptInterface
    fun storeException(url: String, data: String) {
//        AmigoCloudAPI.storeException(url, data)
        print("onException() -----: " + data)
    }

    @JavascriptInterface
    fun openUrl(url: String) {
        formActivity.runOnUiThread(Runnable {
            val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            formActivity.startActivity(browserIntent)
        })
    }

    @JavascriptInterface
    fun setState(json: String) {
        print("setState() -----: " + json)
    }

    @JavascriptInterface
    fun getFormDescription(datasetId: String, formType: String): String {
        if (formType == "create") {
            val json = formViewState.formDescription
            return json
        }
        return "{}"
    }

    @JavascriptInterface
    fun getBlockHTML(formType: String): String {
        var data = ""
        formViewState.form?.let {
            if (formType == "create_block") {
                data = it.create_block_form
            }
        }
        return data
    }

    @JavascriptInterface
    fun getBlockHTMLWithId(formType: String, datasetId: String): String {
        try {
            return "" //AmigoCloudAPI.getDatasetForm(java.lang.Long.parseLong(datasetId), formType)
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getDatasetInfo(datasetId: String): String {
        return "" //AmigoCloudAPI.getDatasetInfo(java.lang.Long.parseLong(datasetId))
    }

    @JavascriptInterface
    fun getDatasetRecords(datasetId: String, offset: String): String {
        return getDatasetRecordsWithFilter(datasetId, offset, "")
    }

    @JavascriptInterface
    fun getDatasetRecordsWithFilter(datasetId: String, offset: String, filter: String): String {
        try {
            return "" //AmigoCloudAPI.getRecordsJSON(java.lang.Long.parseLong(datasetId), Integer.parseInt(offset), filter, "")
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getDatasetRecordsDistinctColumnsWithFilter(datasetId: String, columns: String, filter: String): String {
        var dsid: Long = 0
        try {
            dsid = java.lang.Long.parseLong(datasetId)
        } catch (e: NumberFormatException) {
            e.printStackTrace()
            return "{}"
        }

        return "" //AmigoCloudAPI.getDistinctRecordsJSON(dsid, 0, filter, "", columns)
    }

    @JavascriptInterface
    fun getDatasetRecordsWithFilterOrderBy(datasetId: String, offset: String, filter: String, order_by: String): String {
        try {
            return "" //AmigoCloudAPI.getRecordsJSON(java.lang.Long.parseLong(datasetId), Integer.parseInt(offset), filter, order_by)
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getDatasetRecordsCountWithId(datasetId: String): String {
        return getDatasetRecordsCountWithIdAndFilter(datasetId, "")
    }

    @JavascriptInterface
    fun getDatasetRecordsCountWithIdAndFilter(datasetId: String, filter: String): String {
        try {
            return "" //AmigoCloudAPI.getRecordsCountJSON(java.lang.Long.parseLong(datasetId), filter)
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getRelatedRecordsCount(originalDatasetId: String, relatedDatasetId: String, sourceAmigoId: String): String {
        try {
            return "" //AmigoCloudAPI.getRelatedRecordsCountJSON(java.lang.Long.parseLong(originalDatasetId), java.lang.Long.parseLong(relatedDatasetId), sourceAmigoId)
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getNewRowWithSourceId(datasetId: String, relatedColumn: String, relatedId: String): String {
        try {
            return "" //AmigoCloudAPI.getNewRecordJSONWithRelationship(java.lang.Long.parseLong(datasetId), relatedColumn, relatedId)
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getNewRow(datasetId: String): String {
        try {
            return "" //AmigoCloudAPI.getNewRecordJSONForDataset(java.lang.Long.parseLong(datasetId))
        } catch (e: NumberFormatException) {
            return ""
        }

    }

    @JavascriptInterface
    fun getSchemaWithId(datasetId: String): String {
        return formViewState.schemaJson
    }

    @JavascriptInterface
    fun getSchema(): String
    {
        return formViewState.schemaJson
    }

    @JavascriptInterface
    fun getUser(): String
    {
        return formViewState.userJson
    }

    @JavascriptInterface
    fun getProject(): String
    {
        return formViewState.projectJson
    }

    @JavascriptInterface
    fun getRelatedTables(datasetId: String): String {
        return formViewState.relatedTablesJson
    }

    @JavascriptInterface
    fun getGPSinfo(): String {
        return formActivity.getViewModel().getGPSInfoJSON(lastLocation)
    }

    @JavascriptInterface
    fun getPermissionLevel(): String {
        var perm = "READER"
        formViewState.project?.let {
            perm = it.permission_level
        }
        return perm
    }

    @JavascriptInterface
    fun editRowGeometry(json: String) {
        setGeometryPosition()
    }

    @JavascriptInterface
    fun updateRow(originalJson: String, updateJson: String) {
        formActivity.addNewRecord(updateJson)
    }

    @JavascriptInterface
    fun saveRow(json: String) {
        updateRow(formViewState.recordJson, json)
    }

    @JavascriptInterface
    fun close() {
    }

    @JavascriptInterface
    fun deleteRows(json: String) {}

    @JavascriptInterface
    fun newRecord() {
        print("newRecord")
//        val newRecordData = AmigoCloudAPI.getNewRecordJSONForDataset(datasetId)
//        if (activity is GlobeActivity) {
//            GlobeActivityFragmentExtensionsKt.showEditForm(activity as GlobeActivity, datasetId, newRecordData, true)
//        }
    }

    @JavascriptInterface
    fun zoomToRows(json: String) {}

    @JavascriptInterface
    fun takePhoto(relatedTableId: String , amigoId: String )
    {
        print("$relatedTableId")
        formActivity.takePhoto(relatedTableId, amigoId)
    }

    @JavascriptInterface
    fun scanBarcode(amigoId: String) {
        formActivity.scanBarcode(amigoId)
    }

    fun setCustomFieldValue(fieldName: String, fieldValue: String) {
        runJS("javascript:Amigo.setCustomFieldValue(" +
                formViewState.dataset?.id.toString() + ",'" + fieldName + "','" + fieldValue + "')")
    }

    @JavascriptInterface
    fun viewPhotos(relatedTableId: String, amigoId: String) {
        print("$relatedTableId")
    }

    @JavascriptInterface
    fun dataHasChanged(hasChanged: Boolean) {
        back()
    }

    @JavascriptInterface
    fun writeRfidInfo(lat: String?, lon: String?, owner: String?, asset_name: String?) {
    }

    @JavascriptInterface
    fun getGeometryInfo(wkb: String): String {
        var centroid = "{\"centroid_latitude\":${lastLocation.latitude},"
        centroid += "\"centroid_longitude\":${lastLocation.longitude}}"
        return centroid
    }

    fun mediaAdded() {
        runJS("javascript:Amigo.mediaAdded()")
    }
}